File: packages/worker/src/execution/json-value.ts
import type { JsonValue } from "@tenvyr/contracts";

export function asJsonValue(
  value: unknown,
  path = "$",
  seen = new Set<object>(),
): JsonValue {
  if (value === null || typeof value === "string" || typeof value === "boolean")
    return value;
  if (typeof value === "number") {
    if (!Number.isFinite(value))
      throw new TypeError(`${path} must be a finite JSON number`);
    return value;
  }
  if (typeof value !== "object")
    throw new TypeError(`${path} is not JSON-compatible`);
  const prototype = Object.getPrototypeOf(value);
  if (
    prototype !== Object.prototype &&
    prototype !== null &&
    !Array.isArray(value)
  ) {
    throw new TypeError(`${path} must be a plain JSON object or array`);
  }
  if (seen.has(value))
    throw new TypeError(`${path} contains a circular reference`);
  seen.add(value);
  try {
    if (Array.isArray(value))
      return value.map((item, index) =>
        asJsonValue(item, `${path}[${index}]`, seen),
      );
    const output = Object.create(null) as Record<string, JsonValue>;
    for (const [key, item] of Object.entries(value)) {
      output[key] = asJsonValue(item, `${path}.${key}`, seen);
    }
    return output;
  } finally {
    seen.delete(value);
  }
}

File: sdks/python-worker/src/tenvyr_worker/_protocol/json_value.py
from __future__ import annotations

import math
from typing import TypeAlias, cast

JsonPrimitive: TypeAlias = bool | int | float | str | None
JsonValue: TypeAlias = JsonPrimitive | list["JsonValue"] | dict[str, "JsonValue"]
MAX_SAFE_INTEGER = 9_007_199_254_740_991


class JsonCompatibilityError(TypeError):
    def __init__(self, path: str, message: str, keyword: str) -> None:
        super().__init__(f"{path} {message}")
        self.path = path
        self.message = message
        self.keyword = keyword


def to_json_value(
    value: object, *, _path: str = "$", _seen: set[int] | None = None
) -> JsonValue:
    if value is None or isinstance(value, (str, bool)):
        return value
    if type(value) in (int, float):
        number = cast(int | float, value)
        if isinstance(number, int) and abs(number) > MAX_SAFE_INTEGER:
            raise JsonCompatibilityError(
                _path,
                "integer must be within the interoperable safe range",
                "safeInteger",
            )
        if isinstance(number, float) and not math.isfinite(number):
            raise JsonCompatibilityError(
                _path, "must be a finite JSON number", "finiteNumber"
            )
        if (
            isinstance(number, float)
            and number.is_integer()
            and abs(number) > MAX_SAFE_INTEGER
        ):
            raise JsonCompatibilityError(
                _path,
                "integer must be within the interoperable safe range",
                "safeInteger",
            )
        return number
    if type(value) not in (list, dict):
        raise JsonCompatibilityError(_path, "is not JSON-compatible", "type")

    seen = _seen if _seen is not None else set()
    identity = id(value)
    if identity in seen:
        raise JsonCompatibilityError(_path, "contains a circular reference", "acyclic")
    seen.add(identity)
    try:
        if isinstance(value, list):
            return [
                to_json_value(item, _path=f"{_path}[{index}]", _seen=seen)
                for index, item in enumerate(value)
            ]
        mapping = cast(dict[object, object], value)
        converted: dict[str, JsonValue] = {}
        for key, item in mapping.items():
            if not isinstance(key, str):
                raise JsonCompatibilityError(
                    _path, "must contain only string object keys", "objectKey"
                )
            converted[key] = to_json_value(
                item, _path=_child_path(_path, key), _seen=seen
            )
        return converted
    finally:
        seen.remove(identity)


def _child_path(parent: str, key: str) -> str:
    return f"{parent}.{key}" if key.isidentifier() else f"{parent}[{key!r}]"
